### PR TITLE
 Replace `dynamic` with `object` when substituting constraints.

### LINKED ISSUE #36276: Inconsistent handling of `dynamic` generic type constraint when overriding is involved.
```
#nullable enable
class TestDynamic1<T>
{
    public virtual void M<S>() where S : T
    {
    }
}

class TestDynamic2 : TestDynamic1<dynamic>
{
    public override void M<S>()
    {
    }

    void Test()
    {
        base.M<object?>();
        this.M<object?>();
    }
}
```

Observed:
```
.cs(17,9): warning CS8631: The type 'object?' cannot be used as type parameter 'S' in the generic type or method 'TestDynamic1<dynamic>.M<S>()'. Nullability of type argument 'object?' doesn't match constraint type 'dynamic'.
```

Expected to get an additional similar warning for the ```this.M<object?>()``` method ```TestDynamic2.M<S>()```. Both methods have the same signatures and constraints.

I think this is caused by the fact that 
```
        public static TypeParameterBounds ResolveBounds(
            this TypeParameterSymbol typeParameter,
            AssemblySymbol corLibrary,
            ConsList<TypeParameterSymbol> inProgress,
            ImmutableArray<TypeWithAnnotations> constraintTypes,
            bool inherited,
            CSharpCompilation currentCompilation,
            ArrayBuilder<TypeParameterDiagnosticInfo> diagnosticsBuilder,
            ref ArrayBuilder<TypeParameterDiagnosticInfo> useSiteDiagnosticsBuilder)
```
in ConstraintsHelper filters out `dynamic` constraint. That method shouldn't filter out any constraints. The `dynamic` constraint should be filtered out or converted to `object` in the same places where we decide if `object` constraint should be filtered out or not.

### PR DESCRIPTION
**Pull request by AlekseyTs**
Fixes #36276.

### CODE REVIEW

**Comment by jcouv** [comment]
>                 Assert.Equal("void Test1<dynamic>.M1<S>()", baseM1.ToDisplayString(SymbolDisplayFormat.TestFormatWithConstraints));

`Object~` constraints are omitted, so that's why we don't see a `where S : Object~` constraint here. Is that right?

---
Refers to: src/Compilers/CSharp/Test/Semantic/Semantics/NullableReferenceTypesTests.cs:70292 in 2a68337. [](commit_id = 2a683375dbbebf5c250f453760205e9e8804772f, deletion_comment = False)

File: src/Compilers/CSharp/Portable/Symbols/AbstractTypeMap.cs [hunk 1]
```diff
@@ -326,11 +326,14 @@ internal ImmutableArray<TypeWithAnnotations> SubstituteTypes(ImmutableArray<Type
         /// <summary>
         /// Substitute types, and return the results without duplicates, preserving the original order.
```
Comment (jcouv) [inline src/Compilers/CSharp/Portable/Symbols/AbstractTypeMap.cs]: >/// Substitute types, and return the results without duplicates, preserving the original order. [](start = 8, length = 95)

May be worth adding a comment regarding dynamic.

**Comment by AlekseyTs** [comment]
>                 Assert.Equal("void Test1<dynamic>.M1<S>()", baseM1.ToDisplayString(SymbolDisplayFormat.TestFormatWithConstraints));

> Object~ constraints are omitted, so that's why we don't see a where S : Object~ constraint here. Is that right?

That is correct. Insignificant Object constraints are dropped.

---
In reply to: [501489776](https://github.com/dotnet/roslyn/pull/36379#issuecomment-501489776) [](ancestors = 501489776)

---
Refers to: src/Compilers/CSharp/Test/Semantic/Semantics/NullableReferenceTypesTests.cs:70292 in 2a68337. [](commit_id = 2a683375dbbebf5c250f453760205e9e8804772f, deletion_comment = False)

**Comment by AlekseyTs** [comment]
@cston Please review the last commit.

**Comment by AlekseyTs** [comment]
@dotnet/roslyn-compiler, @cston Please review, need a second sign-off.

**Comment by gafter** [comment]
Is this a change that reflects something that is or should be in the specification?

**Comment by AlekseyTs** [comment]
> Is this a change that reflects something that is or should be in the specification?

I do not think so.

**Comment by AlekseyTs** [comment]
@dotnet/roslyn-compiler, @cston Please review, need a second sign-off.

**Pull request merged**